### PR TITLE
Server Development Merged NKS Code

### PR DESCRIPTION
**Pull request by zhangwei1984**
The Latest Update

Merge all the server-dev branch

Need Merge~

### CODE REVIEW

**Comment by hooohu** [comment]
This is the work from Server Group. We merged the code to handle request.

The code of server behavior should be modified under the condition (flag_test == 0).

The conflicts are resolved. But the request handle is not tested with the client.

**Comment by hooohu** [comment]
Hi,

Our group member has tested the request handle part and find there is a segmentation fault in handling part.

The test without handling request is in file /test/server-kv-test.c and works well.

So, there may be some issues in the handle_request.

**Comment by phil-lopreiato** [comment]
1) Why are there commits from the client side included here? Are the needed?
2) The tests have failed. Please update this pull request so that they pass.

File: lib/shared.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#include <stdlib.h>
```
Comment (phil-lopreiato) [inline lib/shared.h:1]: This header is exactly the same as `marshal.h`. Why is it needed? Seems like it's just adding clutter. If you really think these definitions should be global, mode them to `common.h` in the root directory after merging in https://github.com/gwAdvNet2015/gw-kv-store/tree/server-verbose

File: lib/marshal/marshal.h [hunk 2]
```diff
@@ -1,28 +1,9 @@
...
 #define __MARSHAL_H__
 
-#include <stdlib.h>
-
-/* Defines for returned status */
-#define STORED          0
-#define NOT_STORED      1
-#define EXISTS          2
-#define NOT_FOUND       3
+#include "../shared.h"
```
Comment (phil-lopreiato) [inline lib/marshal/marshal.h:4]: Why are you doing this?

File: client/client-kv.c [hunk 3]
```diff
@@ -0,0 +1,215 @@
+#include <stdlib.h>
```
Comment (phil-lopreiato) [inline client/client-kv.c:1]: Why is this file included here? You're submitting server code. This is made obsolete by #58

File: server/server-kv-run.c [hunk 4]
```diff
@@ -20,14 +20,24 @@
         http://beej.us/guide/bgnet/
 ****************************************/
 
+extern int flag_test;
+
+void
+process_server(char *message)
```
Comment (phil-lopreiato) [inline server/server-kv-run.c:26]: Why is this function necessary? It only calls one thing....

File: server/server-kv.c [hunk 5]
```diff
@@ -20,11 +21,56 @@
...
+
 struct pool_list *list_head = NULL, *list_tail = NULL;
 
 pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
 pthread_cond_t cond = PTHREAD_COND_INITIALIZER;
 pthread_t threads[MAX_CONCURRENCY];
+struct gwkv_server* server;
+
+void
+process_operation(struct gwkv_server *server, struct operation *op, char *message)
```
Comment (phil-lopreiato) [inline server/server-kv.c:36]: I'm pretty sure @nks5295 has already written code to do this in #49. What's the purpose of this?

File: server/server-kv.c [hunk 6]
```diff
@@ -46,21 +92,26 @@ handle_request(void *ptr)
...
-		}
-		else if (bytes_read == 0){
-			printf("Client disconnected");
-		}
-		else {
-                        strcpy(message, gwkv_handle_operation(ht, message));
-                        bytes_write = strlen(message)+1;
-			write(clientfd, message, bytes_write);
-                        printf("thread: %d, received:%s\n", pthread_self(), message);
+		while(1){
```
Comment (phil-lopreiato) [inline server/server-kv.c:95]: What's the value of adding an infinite loop? Shouldn't this be handled per thread?

File: server/server-kv.c [hunk 7]
```diff
@@ -89,6 +140,9 @@ server_main(int sockfd, char* thread_number)
                 addr_size = sizeof client_addr;
                 clientfd = accept(sockfd, (struct sockaddr *)&client_addr, &addr_size);
                 //sh_print_client_ip(client_addr);
+		#ifdef DEBUG
```
Comment (phil-lopreiato) [inline server/server-kv.c:143]: There is already a global flag for verbose printing in the works. See d093778c03580864141f29e48adefa6c509c8e3d

File: server/test/Makefile [hunk 8]
```diff
@@ -1,19 +1,30 @@
 CC = gcc
-CFLAGS = -std=c99 -Wall -pedantic -g
```
Comment (phil-lopreiato) [inline server/test/Makefile:2]: Why are you removing `-Wall` and `-pendantic`? If the code doesn't compile with them, fix the code instead of removing the flags.

File: server/test/Makefile [hunk 9]
```diff
@@ -1,19 +1,30 @@
...
+TARGET = test-ht-wrapper client-kv-test server-kv-test
 
 .PHONY: all test
 
-all: ht_wrapper_test
+all: $(TARGET) 
 
 rebuild: clean all
 
-ht_wrapper_test:
```
Comment (phil-lopreiato) [inline server/test/Makefile:12]: Leave this. This is why the tests are broken.

File: server/test/client-kv-test.c [hunk 10]
```diff
@@ -7,7 +7,7 @@
 #include <unistd.h>
```
Comment (phil-lopreiato) [inline server/test/client-kv-test.c:7]: This code _should_ be implemented by #58.  Is there a reason it's here?

**Comment by phil-lopreiato** [comment]
Lots of things that need to be fixed here. @zhangwei1984 @hantasm  any comments?

**Comment by hooohu** [comment]
Hello, thanks for your code review.

This pull request is merged from various repositories. We worked hard to make sure that the compiling is working.

I agree that there are still lots of issues need to clean up. I will discuss with my group members and then push a new commit.

@phil-lopreiato @zhangwei1984 @wenhuizhang

**Comment by phil-lopreiato** [comment]
Can you go through and reply to my comments above? Some of this pull requests includes things that might be out of your group's scope, so I don't want you do have to do extra work. Also take note of some of the things that you shouldn't have removed (specifically https://github.com/gwAdvNet2015/gw-kv-store/pull/63#discussion_r29203973 and https://github.com/gwAdvNet2015/gw-kv-store/pull/63#discussion_r29203956)

File: lib/shared.h (hunk 1 above)
Comment (wenhuizhang) [inline lib/shared.h:1]: Hi, you are right, we will add "marshal.h" rather than "shared.h" to server site (server-kv & its run/test cases) and eliminate "shared.h"

File: lib/marshal/marshal.h (hunk 2 above)
Comment (wenhuizhang) [inline lib/marshal/marshal.h:4]: Sorry for the modifying your code at marshal.h, as it showed up as a redefined compiling error with shared.h, we mistakenly changed marshal.h, instead of deleting shared.h, and we will include marshal.h later. please stash, and sorry for bothering

File: client/client-kv.c (hunk 3 above)
Comment (wenhuizhang) [inline client/client-kv.c:1]: Oh, could not remember clearly, we used this client for testing server-kv lib, and it showed up as size_t could not be found, that is why <stdlib.h> is included here~~~pls correct if I made it wrong, as actually, stdio.h should have already included defining of "size_t" ~~~

File: server/server-kv-run.c (hunk 4 above)
Comment (wenhuizhang) [inline server/server-kv-run.c:26]: The idea behind this implementation is that, separate server lib (server-kv.c & server-kv.h, which is only a lib without main func) and server internal test (server-kv-run.c which has main func) and the examples to run on server (in folder test named as server-kv-test), to make them as separate blocks, this flag here is to show if it is a internal test for server or run some other external examples/applications/test on the server.

File: server/server-kv.c (hunk 5 above)
Comment (wenhuizhang) [inline server/server-kv.c:36]: As at that time we did not merge nil's code in, this API "process_operation" here is only for temp testing only.

File: server/server-kv.c (hunk 6 above)
Comment (wenhuizhang) [inline server/server-kv.c:95]: It is not a infinite loop, it stops at the point that "curr_char == '\n'";
pls see several lines downward, there is a "if (curr_char == '\n') break;" ,  it jumps out of the loop when "\n" occurs. 

Using similar pattern idea of python loop, such as "for item in lists: "

File: server/server-kv.c (hunk 7 above)
Comment (wenhuizhang) [inline server/server-kv.c:143]: Thank you so much for reminding, we will update "#ifdef DEBUG" to  "#ifdef (verbose_print == 0)"

File: server/test/Makefile (hunk 8 above)
Comment (wenhuizhang) [inline server/test/Makefile:2]: maybe removed in panic ???? Yep, we should have leaved as it is as below
"+CFLAGS = -std=c99 -Wall -pedantic -g -pthread"

File: server/test/Makefile (hunk 9 above)
Comment (wenhuizhang) [inline server/test/Makefile:12]: We updated and put them together in the test session: 

"+test:
-   -./test-ht-wrapper
-   ./server-kv-test &
-   sleep 1
-   ./client-kv-test
-   killall server-kv-test
  "

File: server/test/client-kv-test.c (hunk 10 above)
Comment (wenhuizhang) [inline server/test/client-kv-test.c:7]: For testing as an application for server, we need a client....

File: server/server-kv.c (hunk 7 above)
Comment (phil-lopreiato) [inline server/server-kv.c:143]: It's even easier, actually. Merge in the branch I linked and use the function `kvprintf()` which will only print if the verbose flag is set

File: server/test/Makefile (hunk 9 above)
Comment (phil-lopreiato) [inline server/test/Makefile:12]: Ah, gotcha. Can you leave the `ht-wrapper` tests separate though? It's beneficial to run `make ht_wrapper_test` to just run a subset of tests

File: server/test/client-kv-test.c (hunk 10 above)
Comment (phil-lopreiato) [inline server/test/client-kv-test.c:7]: It might be best to wait until #58 is ready before we merge this then, in order to avoid duplicates.

File: server/server-kv.c (hunk 7 above)
Comment (wenhuizhang) [inline server/server-kv.c:143]: Sounds great ! @zhangwei1984 @hantasm

File: server/test/Makefile (hunk 9 above)
Comment (wenhuizhang) [inline server/test/Makefile:12]: Sure, as you wish :) @zhangwei1984  @hantasm

File: server/test/client-kv-test.c (hunk 10 above)
Comment (wenhuizhang) [inline server/test/client-kv-test.c:7]: Good point @zhangwei1984 @hantasm  let it go ~~~ Just Joking

**Comment by amygara** [comment]
Closed due to updates: see PRs #65 and #66
